src/buildercore/cfngen.py

```diff
@@ -109,108 +120,169 @@ def build_context(pname, **more_context): # pylint: disable=too-many-locals
+        build_context_elasticache,
+        build_context_vault,
+    ]
+
+    # exceptions to the rule ...
+    wrangler_list = [project_wrangler] + wrangler_list
+
+    for wrangler in wrangler_list:
+        # so functions can't modify the context in-place or
+        # reference a bit of project_data and then change it
+        context = wrangler(deepcopy(project_data), deepcopy(context))
+
+    return context
+
+def build_context_aws(pdata, context):
+    if 'aws' not in pdata:
+        return context
+    keepers = [
+        'region',
+        'vpc-id',
+        'subnet-id',
+        'subnet-cidr',
+        'redundant-subnet-id',
+        'redundant-subnet-cidr',
+    ]
+    context['aws'] = subdict(pdata['aws'], keepers)
+    return context
+
+def build_context_s3(pdata, context):
+    # future: build what is necessary for buildercore.bootstrap.setup_s3()
+    default_bucket_configuration = {
+        'sqs-notifications': {},
+        'deletion-policy': 'delete',
+        'website-configuration': None,
+        'cors': None,
+        'public': False,
+    }
+    for bucket_template_name in pdata['aws'].get('s3', {}):
+        bucket_name = parameterize(context)(bucket_template_name)
+        configuration = pdata['aws']['s3'][bucket_template_name]
+        context['s3'][bucket_name] = default_bucket_configuration.copy()
+        context['s3'][bucket_name].update(configuration if configuration else {})
+    return context
 
-    # stackname data
-    bits = core.parse_stackname(stackname, all_bits=True, idx=True)
+def build_context_sns_sqs(pdata, context):
+    _parameterize = parameterize(context)
+
+    for topic_template_name in pdata['aws'].get('sns', []):
+        topic_name = _parameterize(topic_template_name)
+        context['sns'].append(topic_name)
+
+    for queue_template_name in pdata['aws'].get('sqs', {}):
+        queue_name = _parameterize(queue_template_name)
+        queue_configuration = pdata['aws']['sqs'][queue_template_name]
+        subscriptions = lmap(_parameterize, queue_configuration.get('subscriptions', []))
+        context['sqs'][queue_name] = subscriptions
+    return context
+
+def project_wrangler(pdata, context):
+    bits = core.parse_stackname(context['stackname'], all_bits=True, idx=True)
+    pname = context['project_name']
     ensure(bits['project_name'] == pname,
            "the project name %r derived from the given `stackname` %r doesn't match" % (bits['project_name'], pname))
+    # provides 'project_name', 'instance_id', 'cluster_id'
     context.update(bits)
 
     # hostname data
-    context.update(core.hostname_struct(stackname))
+    # provides: 'domain', 'int_domain', 'subdomain',
+    #           'hostname', 'project_hostname', 'int_project_hostname',
+    #           'full_hostname', 'int_full_hostname'
+    context.update(core.hostname_struct(context['stackname']))
+
+    # project data
+    # preseve some of the project data. all of it is too much
+    keepers = [
+        'salt',
```

Comment:
later on `salt` should probably be moved to `ec2` as it's only used for that. `formula-*` fields can in theory be used with K8S too so it's the right place.

Reply:
this snippet was copied from trop.py and trop.py pares down even further